rpctypes/types.go

```diff
@@ -211,19 +233,114 @@ func (b *EthSendBundleArgs) Validate() (common.Hash, uuid.UUID, error) {
-	var buf []byte
-	buf = binary.AppendVarint(buf, int64(blockNumber))
-	buf = append(buf, hashBytes...)
-	sort.Slice(b.RevertingTxHashes, func(i, j int) bool {
-		return bytes.Compare(b.RevertingTxHashes[i][:], b.RevertingTxHashes[j][:]) <= 0
-	})
-	for _, txHash := range b.RevertingTxHashes {
-		buf = append(buf, txHash[:]...)
+	if b.Version == nil || *b.Version == BundleVersionV1 {
+		// then compute the uuid
+		var buf []byte
+		buf = binary.AppendVarint(buf, int64(blockNumber))
+		buf = append(buf, hashBytes...)
+		sort.Slice(b.RevertingTxHashes, func(i, j int) bool {
+			return bytes.Compare(b.RevertingTxHashes[i][:], b.RevertingTxHashes[j][:]) <= 0
+		})
+		for _, txHash := range b.RevertingTxHashes {
+			buf = append(buf, txHash[:]...)
+		}
+		return common.BytesToHash(hashBytes),
+			uuid.NewHash(sha256.New(), uuid.Nil, buf, 5),
+			nil
 	}
-	return common.BytesToHash(hashBytes),
-		uuid.NewHash(sha256.New(), uuid.Nil, buf, 5),
-		nil
+
+	if *b.Version == BundleVersionV2 {
+		// blockNumber, default 0
+		blockNumber := uint64(0)
+		if b.BlockNumber != nil {
+			blockNumber = uint64(*b.BlockNumber)
+		}
+
+		// minTimestamp, default 0
+		minTimestamp := uint64(0)
+		if b.MinTimestamp != nil {
+			minTimestamp = *b.MinTimestamp
+		}
+
+		// maxTimestamp, default ^uint64(0) (i.e. 0xFFFFFFFFFFFFFFFF in Rust)
+		maxTimestamp := ^uint64(0)
+		if b.MaxTimestamp != nil {
+			maxTimestamp = *b.MaxTimestamp
+		}
+
+		// Build up our buffer using variable-length encoding of the block
+		// number, minTimestamp, maxTimestamp, #revertingTxHashes, #droppingTxHashes.
+		var buf []byte
+		buf = binary.AppendUvarint(buf, blockNumber)
+		buf = binary.AppendUvarint(buf, minTimestamp)
+		buf = binary.AppendUvarint(buf, maxTimestamp)
+		buf = binary.AppendUvarint(buf, uint64(len(b.RevertingTxHashes)))
+		buf = binary.AppendUvarint(buf, uint64(len(b.DroppingTxHashes)))
+
+		// Append the main txs keccak hash (already computed in hashBytes).
+		buf = append(buf, hashBytes...)
+
+		// Sort revertingTxHashes and append them.
+		sort.Slice(b.RevertingTxHashes, func(i, j int) bool {
+			return bytes.Compare(b.RevertingTxHashes[i][:], b.RevertingTxHashes[j][:]) < 0
+		})
+		for _, h := range b.RevertingTxHashes {
+			buf = append(buf, h[:]...)
+		}
+
+		// Sort droppingTxHashes and append them.
+		sort.Slice(b.DroppingTxHashes, func(i, j int) bool {
+			return bytes.Compare(b.DroppingTxHashes[i][:], b.DroppingTxHashes[j][:]) < 0
+		})
+		for _, h := range b.DroppingTxHashes {
+			buf = append(buf, h[:]...)
+		}
+
+		// If a "refund" is present (analogous to the Rust code), we push:
+		//   refundPercent (1 byte)
+		//   refundRecipient (20 bytes, if an Ethereum address)
+		//   #refundTxHashes (varint)
+		//   each refundTxHash (32 bytes)
+		// NOTE: The Rust code uses a single byte for `refund.percent`,
+		//       so we do the same here. Adjust as needed if your data differs.
+		if b.RefundPercent != nil && b.RefundRecipient != nil && len(b.RefundTxHashes) > 0 {
```

Comment:
Is it && or || here?

I think that we can have refund percent set but anything else can be unset.

Also, add tests here + ask DanielX about tests for this in rbuilder. Because I am looking at rbuilder code right now and it looks like it always sets recipient and tx_hashes because we parse txs and here we don't parse txs.